In [0]:
import numpy as np
from sklearn import datasets
from matplotlib import pyplot as plt

In [0]:
boston = datasets.load_boston()

In [5]:
type(boston)
boston

{'DESCR': ".. _boston_dataset:\n\nBoston house prices dataset\n---------------------------\n\n**Data Set Characteristics:**  \n\n    :Number of Instances: 506 \n\n    :Number of Attributes: 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target.\n\n    :Attribute Information (in order):\n        - CRIM     per capita crime rate by town\n        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.\n        - INDUS    proportion of non-retail business acres per town\n        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)\n        - NOX      nitric oxides concentration (parts per 10 million)\n        - RM       average number of rooms per dwelling\n        - AGE      proportion of owner-occupied units built prior to 1940\n        - DIS      weighted distances to five Boston employment centres\n        - RAD      index of accessibility to radial highways\n        - TAX      full-value property-tax rate p

In [6]:
boston.keys()

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename'])

In [7]:
print(boston["DESCR"])

.. _boston_dataset:

Boston house prices dataset
---------------------------

**Data Set Characteristics:**  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target.

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index of accessibility to radial highways
        - TAX      full-value property-tax rate per $10,000
        - PTRATIO  pu

In [8]:
boston['data']

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, ..., 1.5300e+01, 3.9690e+02,
        4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9690e+02,
        9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9283e+02,
        4.0300e+00],
       ...,
       [6.0760e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        5.6400e+00],
       [1.0959e-01, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9345e+02,
        6.4800e+00],
       [4.7410e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        7.8800e+00]])

In [9]:
boston['target']

array([24. , 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9, 15. ,
       18.9, 21.7, 20.4, 18.2, 19.9, 23.1, 17.5, 20.2, 18.2, 13.6, 19.6,
       15.2, 14.5, 15.6, 13.9, 16.6, 14.8, 18.4, 21. , 12.7, 14.5, 13.2,
       13.1, 13.5, 18.9, 20. , 21. , 24.7, 30.8, 34.9, 26.6, 25.3, 24.7,
       21.2, 19.3, 20. , 16.6, 14.4, 19.4, 19.7, 20.5, 25. , 23.4, 18.9,
       35.4, 24.7, 31.6, 23.3, 19.6, 18.7, 16. , 22.2, 25. , 33. , 23.5,
       19.4, 22. , 17.4, 20.9, 24.2, 21.7, 22.8, 23.4, 24.1, 21.4, 20. ,
       20.8, 21.2, 20.3, 28. , 23.9, 24.8, 22.9, 23.9, 26.6, 22.5, 22.2,
       23.6, 28.7, 22.6, 22. , 22.9, 25. , 20.6, 28.4, 21.4, 38.7, 43.8,
       33.2, 27.5, 26.5, 18.6, 19.3, 20.1, 19.5, 19.5, 20.4, 19.8, 19.4,
       21.7, 22.8, 18.8, 18.7, 18.5, 18.3, 21.2, 19.2, 20.4, 19.3, 22. ,
       20.3, 20.5, 17.3, 18.8, 21.4, 15.7, 16.2, 18. , 14.3, 19.2, 19.6,
       23. , 18.4, 15.6, 18.1, 17.4, 17.1, 13.3, 17.8, 14. , 14.4, 13.4,
       15.6, 11.8, 13.8, 15.6, 14.6, 17.8, 15.4, 21

In [10]:
boston['feature_names'] # 열이름

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

- CRIM: 범죄율
- INDUS: 비소매상업지역 면적 비율
- NOX: 일산화질소 농도
- RM: 주택당 방 수
- LSTAT: 인구 중 하위 계층 비율
- B: 인구 중 흑인 비율
- PTRATIO: 학생/교사 비율
- ZN: 25,000 평방피트를 초과 거주지역 비율
- CHAS: 찰스강의 경계에 위치한 경우는 1, 아니면 0
- AGE: 1940년 이전에 건축된 주택의 비율
- RAD: 방사형 고속도로까지의 거리
- DIS: 직업센터의 거리
- TAX: 재산세율
- MEDV   보스턴 506개 타운의 1978년 주택 가격 중앙값 (단위 1,000 달러)


    

In [11]:
nd_boston= np.array(boston)
type(nd_boston)

numpy.ndarray

In [0]:
nd_boston_data = np.array(boston['data'],dtype= np.int64)
nd_boston_target = np.array(boston['target'])

In [13]:
nd_boston_data

array([[  0,  18,   2, ...,  15, 396,   4],
       [  0,   0,   7, ...,  17, 396,   9],
       [  0,   0,   7, ...,  17, 392,   4],
       ...,
       [  0,   0,  11, ...,  21, 396,   5],
       [  0,   0,  11, ...,  21, 393,   6],
       [  0,   0,  11, ...,  21, 396,   7]])

In [0]:
crim =[nd_boston_data[:,0]]
zn = [nd_boston_data[:,1]]
indus = [nd_boston_data[:,2]]
chas = [nd_boston_data[:,3]]
nox = [nd_boston_data[:,4]]
rm = [nd_boston_data[:,5]]
age = [nd_boston_data[:,6]]
dis = [nd_boston_data[:,7]]
rad = [nd_boston_data[:,8]]
tax = [nd_boston_data[:,9]]
ptratio = [nd_boston_data[:,10]]
b = [nd_boston_data[:,11]]
lstat = [nd_boston_data[:,12]]

In [15]:
print(nd_boston_data)

[[  0  18   2 ...  15 396   4]
 [  0   0   7 ...  17 396   9]
 [  0   0   7 ...  17 392   4]
 ...
 [  0   0  11 ...  21 396   5]
 [  0   0  11 ...  21 393   6]
 [  0   0  11 ...  21 396   7]]


In [16]:
boston_mean =[]  
boston_median = []
boston_std = []
boston_var=[]

for i in range(13):
  boston_mean +=[round(np.mean(nd_boston_data[:,i]),2)]  
  boston_median += [np.median(nd_boston_data[:,i])]
  boston_std += [round(np.std(nd_boston_data[:,i]),2)]
  boston_var += [round(np.var(nd_boston_data[:,i]),2)]
  print(i,"번째 min 값 :",np.min(nd_boston_data[:,i]))
  print(i,"번째 max 값 :",np.max(nd_boston_data[:,i]))
  print()

0 번째 min 값 : 0
0 번째 max 값 : 88

1 번째 min 값 : 0
1 번째 max 값 : 100

2 번째 min 값 : 0
2 번째 max 값 : 27

3 번째 min 값 : 0
3 번째 max 값 : 1

4 번째 min 값 : 0
4 번째 max 값 : 0

5 번째 min 값 : 3
5 번째 max 값 : 8

6 번째 min 값 : 2
6 번째 max 값 : 100

7 번째 min 값 : 1
7 번째 max 값 : 12

8 번째 min 값 : 1
8 번째 max 값 : 24

9 번째 min 값 : 187
9 번째 max 값 : 711

10 번째 min 값 : 12
10 번째 max 값 : 22

11 번째 min 값 : 0
11 번째 max 값 : 396

12 번째 min 값 : 1
12 번째 max 값 : 37



In [17]:
print(boston_mean)
print(boston_median)
print(boston_std)
print(boston_var)

[3.32, 11.35, 10.71, 0.07, 0.0, 5.78, 68.16, 3.29, 9.55, 408.24, 18.08, 356.08, 12.17]
[0.0, 0.0, 9.0, 0.0, 0.0, 6.0, 77.0, 3.0, 5.0, 330.0, 19.0, 391.0, 11.0]
[8.49, 23.29, 6.93, 0.25, 0.0, 0.78, 28.18, 2.14, 8.7, 168.37, 2.28, 91.17, 7.14]
[72.16, 542.31, 47.98, 0.06, 0.0, 0.6, 793.86, 4.57, 75.67, 28348.62, 5.19, 8311.36, 50.98]


- CRIM,ZN,INDUS,AGE,TAX ,B 컬럼의 극단치는 높으므로 대푯값으로 중앙값을 사용하는것이 좋다고 판단됩니다.
- CHAS,NOX,DIS,RAD,PTRATIO,LSTAT 컬럼의 극단치는 낮으므로 대푯값으로 평균을 사용하는게 좋다고 판단됩니다.

##시각화

In [0]:
boston_target = boston['target']

In [19]:
len(boston_target)

506

#판다스로 불러오기

In [0]:
import pandas as pd

In [0]:
se_boston = pd.Series(boston)
df_boston = pd.DataFrame(se_boston)

In [25]:
df_boston.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5 entries, data to filename
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       5 non-null      object
dtypes: object(1)
memory usage: 80.0+ bytes


In [26]:
df_boston.describe()

/usr/local/lib/python3.6/dist-packages/pandas/core/algorithms.py:767: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  keys, counts = f(values, dropna)


TypeError: ignored

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas/_libs/hashtable_class_helper.pxi", line 1653, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'numpy.ndarray'


0
count                                                   5
unique                                                  5
top     [CRIM, ZN, INDUS, CHAS, NOX, RM, AGE, DIS, RAD...
freq                                                    1

In [27]:
df_boston.head()

0
data           [[0.00632, 18.0, 2.31, 0.0, 0.538, 6.575, 65.2...
target         [24.0, 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27....
feature_names  [CRIM, ZN, INDUS, CHAS, NOX, RM, AGE, DIS, RAD...
DESCR          .. _boston_dataset:\n\nBoston house prices dat...
filename       /usr/local/lib/python3.6/dist-packages/sklearn...

In [28]:
df_boston.tail()

0
data           [[0.00632, 18.0, 2.31, 0.0, 0.538, 6.575, 65.2...
target         [24.0, 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27....
feature_names  [CRIM, ZN, INDUS, CHAS, NOX, RM, AGE, DIS, RAD...
DESCR          .. _boston_dataset:\n\nBoston house prices dat...
filename       /usr/local/lib/python3.6/dist-packages/sklearn...